File: Server/logger/core/parse_log_path.py
"""Parse dot-notation log category path inputs."""

from __future__ import annotations

from dataclasses import dataclass

from .categories import LOG_PATHS, LogCategory


@dataclass(frozen=True)
class ParsedLogPath:
    path: str
    category: LogCategory
    subcategory: str | None
    category_label: str


_LOG_PATH_SET = frozenset(LOG_PATHS)
_API_SUBCATEGORIES = frozenset(
    path.split(".", 1)[1] for path in LOG_PATHS if path.startswith("API.")
)


def parse_log_path(value: LogCategory | str) -> ParsedLogPath:
    if isinstance(value, LogCategory):
        normalized = value.value
    else:
        normalized = value.strip()

    if not normalized:
        raise ValueError("Log path must be a non-empty string")

    if normalized in _LOG_PATH_SET:
        if "." not in normalized:
            category = LogCategory(normalized)
            return ParsedLogPath(
                path=normalized,
                category=category,
                subcategory=None,
                category_label=normalized,
            )

        category_part, sub_part = normalized.split(".", 1)
        category = LogCategory(category_part)
        return ParsedLogPath(
            path=normalized,
            category=category,
            subcategory=sub_part,
            category_label=normalized,
        )

    if "." in normalized:
        category_part, sub_part = normalized.split(".", 1)
        if category_part == LogCategory.API.value and sub_part in _API_SUBCATEGORIES:
            category = LogCategory.API
            return ParsedLogPath(
                path=normalized,
                category=category,
                subcategory=sub_part,
                category_label=normalized,
            )

    raise ValueError(f"Unknown log path: {normalized}")

File: Client/packages/logger/core/parseLogPath.ts
import type { ApiSubcategory, LogCategory, LogPath } from "./categories";
import { API_SUBCATEGORIES, LOG_CATEGORIES, LOG_PATHS } from "./categories";

export type ParsedLogPath = {
  path: LogPath | string;
  category: LogCategory;
  subcategory?: ApiSubcategory;
  categoryLabel: string;
};

const LOG_PATH_SET = new Set<string>(LOG_PATHS);

function isLogCategory(value: string): value is LogCategory {
  return value in LOG_CATEGORIES;
}

function isApiSubcategory(value: string): value is ApiSubcategory {
  return value in API_SUBCATEGORIES;
}

export function parseLogPath(input: LogCategory | LogPath | string): ParsedLogPath {
  if (isLogCategory(input)) {
    return {
      path: input,
      category: input,
      categoryLabel: input,
    };
  }

  const normalized = input.trim();
  if (!normalized) {
    throw new Error("Log path must be a non-empty string");
  }

  if (LOG_PATH_SET.has(normalized)) {
    const dotIndex = normalized.indexOf(".");
    if (dotIndex === -1) {
      const category = normalized as LogCategory;
      return {
        path: category,
        category,
        categoryLabel: category,
      };
    }

    const categoryPart = normalized.slice(0, dotIndex);
    const subPart = normalized.slice(dotIndex + 1);
    if (!isLogCategory(categoryPart) || !isApiSubcategory(subPart)) {
      throw new Error(`Invalid log path: ${normalized}`);
    }

    return {
      path: normalized as LogPath,
      category: categoryPart,
      subcategory: subPart,
      categoryLabel: normalized,
    };
  }

  const dotIndex = normalized.indexOf(".");
  if (dotIndex !== -1) {
    const categoryPart = normalized.slice(0, dotIndex);
    const subPart = normalized.slice(dotIndex + 1);
    if (isLogCategory(categoryPart) && isApiSubcategory(subPart)) {
      return {
        path: normalized,
        category: categoryPart,
        subcategory: subPart,
        categoryLabel: normalized,
      };
    }
  }

  if (isLogCategory(normalized)) {
    return {
      path: normalized,
      category: normalized,
      categoryLabel: normalized,
    };
  }

  throw new Error(`Unknown log path: ${normalized}`);
}
